parlai/tasks/multiwoz/agents.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from parlai.core.teachers import FixedDialogTeacher
+from .build import build
+import os
+import json
+
+
+def _path(opt, filtered):
+    # build the data if it does not exist
+    build(opt)
+
+    # set up path to data (specific to each dataset)
+    jsons_path = os.path.join(opt['datapath'], 'multiwoz', 'MULTIWOZ2.1')
+    conversations_path = os.path.join(jsons_path, 'data.json')
+    return conversations_path, jsons_path
+
+
+class MultiWozTeacher(FixedDialogTeacher):
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        opt['datafile'], jsons_path = _path(opt, '')
+        self._setup_data(opt['datafile'], jsons_path)
+        self.id = 'multiwoz'
+        self.reset()
+
+    def _setup_data(self, data_path, jsons_path):
+        print('loading: ' + data_path)
+        with open(data_path) as data_file:
+            self.messages = json.load(data_file)
+
+        test_path = os.path.join(jsons_path, 'testListFile.json')
+        valid_path = os.path.join(jsons_path, 'valListFile.json')
+        if self.datatype.startswith('test'):
+            with open(test_path) as f:
+                test_data = {self.messages[line] for line in f}
+                self.messages = test_data
+        elif self.datatype.startswith('valid'):
+            with open(valid_path) as f:
+                valid_data = {self.messages[line] for line in f}
+                self.messages = valid_data
+        else:
+            with open(test_path) as f:
+                for line in f:
+                    if line in self.messages:
+                        del self.messages[line]
+            with open(valid_path) as f:
+                for line in f:
+                    if line in self.messages:
+                        del self.messages[line]
+        self.messages = list(self.messages.values())
+
+    def num_examples(self):
+        examples = 0
+        for data in self.messages:
+            examples += len(data['log']) // 2
+        return examples
+
+    def num_episodes(self):
+        return len(self.messages)
+
+    def get(self, episode_idx, entry_idx=0):
+        log_idx = entry_idx * 2
```

Comment:
could you write a comment/docstring somewhere that indicates what else is available in the data besides just the dialogue? we may not want/need anything besides the dialogue, but the dataset indeed has a lot more substance than just a message log

Reply:
I added comments in agents.py. Let me know if those look alright or if I should put them somewhere else or add more info.

Reply:
these look good, but i'd prefer it in a docstring rather than a list of comments, e.g. something like this:

```
"""Multiwoz Teacher.

This dataset contains...

...

"""
```